core/src/main/java/org/infinispan/interceptors/impl/CacheWriterInterceptor.java

```diff
@@ -175,43 +183,66 @@ private Transaction suspendRunningTx(TxInvocationContext ctx) throws SystemExcep
 
    @Override
    public CompletableFuture<Void> visitPutKeyValueCommand(InvocationContext ctx, PutKeyValueCommand command) throws Throwable {
-      Object returnValue = ctx.forkInvocationSync(command);
-      if (!isStoreEnabled(command) || ctx.isInTxScope() || !command.isSuccessful()) return ctx.shortCircuit(returnValue);
-      if (!isProperWriter(ctx, command, command.getKey())) return ctx.shortCircuit(returnValue);
+      ctx.onReturn((ctx1, command1, rv, throwable) -> {
+         if (throwable != null)
+            throw throwable;
 
-      Object key = command.getKey();
-      storeEntry(ctx, key, command);
-      if (getStatisticsEnabled()) cacheStores.incrementAndGet();
+         if (!isStoreEnabled(command) || ctx.isInTxScope() || !command.isSuccessful())
+            return null;
+         if (!isProperWriter(ctx, command, command.getKey()))
+            return null;
 
-      return ctx.shortCircuit(returnValue);
+         Object key = command.getKey();
+         storeEntry(ctx, key, command);
+         if (getStatisticsEnabled())
+            cacheStores.incrementAndGet();
+         return null;
+      });
+      return ctx.continueInvocation();
    }
 
    @Override
    public CompletableFuture<Void> visitReplaceCommand(InvocationContext ctx, ReplaceCommand command) throws Throwable {
-      Object returnValue = ctx.forkInvocationSync(command);
-      if (!isStoreEnabled(command) || ctx.isInTxScope() || !command.isSuccessful()) return ctx.shortCircuit(returnValue);
-      if (!isProperWriter(ctx, command, command.getKey())) return ctx.shortCircuit(returnValue);
+      ctx.onReturn((ctx1, command1, rv, throwable) -> {
+         if (throwable != null)
+            throw throwable;
+
+         if (!isStoreEnabled(command) || ctx.isInTxScope() || !command.isSuccessful())
+            return null;
+         if (!isProperWriter(ctx, command, command.getKey()))
+            return null;
 
-      Object key = command.getKey();
-      storeEntry(ctx, key, command);
-      if (getStatisticsEnabled()) cacheStores.incrementAndGet();
+         Object key = command.getKey();
+         storeEntry(ctx, key, command);
+         if (getStatisticsEnabled())
+            cacheStores.incrementAndGet();
+         return null;
+      });
 
-      return ctx.shortCircuit(returnValue);
+      return ctx.continueInvocation();
```

Comment:
Yeah, there was a time when `onReturn` didn't return a `CF<Void>`, and some of the interceptors got stuck in the past...